craft_store/store_client.py

```diff
@@ -0,0 +1,169 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Craft Store StoreClient."""
+
+import base64
+import json
+from typing import TYPE_CHECKING, Any, Dict, Sequence
+from urllib.parse import urlparse
+
+import macaroonbakery._utils as utils
+import requests
+from macaroonbakery import bakery, httpbakery
+
+from .auth import Auth
+from .http_client import HTTPClient
+
+if TYPE_CHECKING:
+    from . import endpoints
+
+
+class StoreClient(HTTPClient):
+    """Encapsulates API calls for the Snap Store or Charmhub."""
+
+    def __init__(
+        self,
+        *,
+        base_url: str,
+        endpoints: "endpoints.Endpoints",
+        application_name: str,
+        user_agent: str,
+    ) -> None:
+        """Initialize the Store Client.
+
+        :param base_url: the base url of the API endpoint.
+        :param endpoints: :data:`endpoints.CHARMHUB` or :data:`endpoints.SNAP_STORE`.
+        :param application_name: the name application using this class, used for the keyring.
+        :param user_agent: User-Agent header to use for HTTP(s) requests.
+        """
+        super().__init__(user_agent=user_agent)
+
+        self._bakery_client = httpbakery.Client()
+        self._base_url = base_url
+        self._store_host = urlparse(base_url).netloc
+        self._endpoints = endpoints
+        self._auth = Auth(application_name, base_url)
+
+    def _get_token(self, token_request: Dict[str, Any]) -> str:
+        token_response = super().request(
+            "POST",
+            self._base_url + self._endpoints.tokens,
+            json=token_request,
+        )
+
+        return token_response.json()["macaroon"]
+
+    def _candid_discharge(self, macaroon: str) -> str:
+        bakery_macaroon = bakery.Macaroon.from_dict(json.loads(macaroon))
+        discharges = bakery.discharge_all(
+            bakery_macaroon, self._bakery_client.acquire_discharge
+        )
+
+        # serialize macaroons the bakery-way
+        discharged_macaroons = (
+            "[" + ",".join(map(utils.macaroon_to_json_string, discharges)) + "]"
```

Comment:
Consider re-implementing `macaroon_to_json_string` so we don't need to use a private module from the lib. The function is actually one line, and it uses "public" helpers.

Reply:
updated in https://github.com/canonical/craft-store/pull/6/commits/e12ac2076348199eff3867e11575cdbdf107d67c